### ISSUE TITLE
Bower packages are case sensitive

### ISSUE THREAD

**Issue by samdark** [francoispluchino/composer-asset-plugin#70]
It seems currently plugin lowercases package name and it seems to be wrong since Bower package names are case sensitive. See https://github.com/yiisoft/yii2/issues/6276

**Comment by neoascetic** [comment]
Similar issue here. [Here](https://github.com/desandro/imagesloaded/blob/master/bower.json#L7) "imagesloaded" bower package specifies "eventEmitter" with version "4.x" as its dependency. [Here](http://bower.io/search/?q=eventEmitter) bower packages for "eventEmitter" query. The plugin tries to use [this package, which has "eventemitter" name](https://github.com/VictorQueiroz/eventemitter/blob/master/package.json) instead of [that one, which is correct dependency](https://github.com/Wolfy87/EventEmitter/blob/master/bower.json) (although it is displayed as "EventEmitter.js" on search results page).

I've wasted 2 hours of my life to find and fix this issue (with repositories overriding). Even if this is wontfix, I am sure there should be some note in FAQ.

Thanks.

**Comment by francoispluchino** [comment]
@neoascetic it's true, this information is not in the FAQ. If you want to add this doc, I'm interested.

**Comment by schmunk42** [comment]
same here

**Comment by philippfrenzel** [comment]
Hey, thanks for the advise! With the new "extra" it works fine! And travis build is back without errors!

**Comment by schmunk42** [comment]
To be honest, this is really annoying and defining repos in the root composer.json are more a hotfix than a solution. But instead of complaining, let me ask some questions ;)

> Problem: the package dependencies are normalized to lowercase on loading the file, therefore, the package name is always lowercase for VCS Repositories.

- Where does this happen? 
- Is it a composer or bower specific thing?
- Could we keep the original name somewhere in the extras section of a generated composer.json for bower repos?
- Is the problem related to the solver, would the solver accept camel-case packages?

Feature request in the meantime: the plugin should at least issue a warning message if two identical (case-insensitive) packages are found/installed.

**Comment by francoispluchino** [comment]
I agree with you, this is just a workaround, not a real solution, but believe me, I tried all possible solutions... But it's very difficult when we must to touch the internal components of Composer. The lead developers of Composer will not touch the Solver SAT, and that the changes will bring an improvement, only for our plugin, which is not necessary for them, given that there are already tools to manage this type of dependency, called: NPM and Bower...

Perhaps now you will understand why I wanted to move on (see proposal v2.0), given the blockage encountered.

Concerning your proposal, it is not possbile, because the `AbstractAssetsRepository` receives the package name to lowercase from the Solver SAT with the method `Fxp\Composer\AssetPlugin\Repository\AbstractAssetsRepository::whatProvides()`. The standardization process of dependency name is carried out during the conversion of the packages to instance `MemoryPackage` via the class [Composer\Package\Link](https://github.com/composer/composer/blob/master/src/Composer/Package/Link.php). The plugin therefore cannot retrieve the original name. It would have been preferable to transmit the instance of the class `Link` to the method `whatProvides()`, in this way, we would have managed to get the original name of dependency.

Another point, I cant detect if there are several version with different case, given that the `AbstractAssetsRepository` receives the lowercase name, and the GET API of Bower and NPM returns a result for a specific value. And execute a new search each time will be an additional blows on performance... So it will be difficult to put a warning. Also, how do you know if the plugin should display a warning or not?

However, we can try to automatically add the configs of VCS, located in `extra.asset-repositories` of the package Composer, but, what behavior must have the plugin in case of duplication? An exception?

This solution requires that the developer to the PHP library include this vcs repository into her `composer.json` file.

**Comment by neoascetic** [comment]
Each bower package has "name" attribute in `bower.json` which may be analyzed instead of lowercased package name from `composer.json`.

**Comment by francoispluchino** [comment]
It's the Solver SAT that determines research in the Repositories. Donc il est impossible (see previous comment) Therefore it is impossible to recover the real name of the package.

**Comment by schmunk42** [comment]
Just thinking loud...

```
eventemitter --> eventemitter
eventEmitter --> event__uc-emitter
EventEmitterX --> __uc-event__uc-emitter__uc-x
```

Any uppercase character gets prepended with `__uc-`. *Could even be made uglier* :)

Would that - just technically - fix this issue?
Could you normalize the name in another way for composer?

PS: I believe you totally about all the rest :smiley:

**Comment by francoispluchino** [comment]
So, for the package dependencies from an asset dependency, yes it is achievable, but in this case, it would be best to keep a readable name (stock maps for alias). But for asset dependencies included in the composer.json file, the name can not be converted, and the problem is still present. Of course, we can indicate in the doc, to format correctly the package name with your pattern, mais sérieusement, it is not legible, and the error message, if any, do not be understandable.

IMO, I would tend to automatically include the repositories in the `extra.asset-repositories`, if possible.

**Comment by philippfrenzel** [comment]
je pense, que tobias fait le proposal pour convertez automatique - comme ca il y a possible de communique avec error message "comprendable" ;)

**Comment by francoispluchino** [comment]
@philippfrenzel ???

**Comment by schmunk42** [comment]
Excusez-moi, quand je ne parle pas français ici :)

So I looked at: https://github.com/bower/bower.json-spec#name - and they don't allow uppercase packages names, but there are some packages registered with UpperCase names.

Am I overlooking something here or is this just pure crap?!

So to me it looks like `[a-zA-Z0-9._-]` are de-facto possible characters for bower packages.
Even more, like `/`? I wouldn't be surprised.

> would be best to keep a readable name

Any suggestions?

> But for asset dependencies included in the composer.json file, the name can not be converted,

It would be just about the name in the file-system. For sure file references may have to be adopted.
btw: How does bower handle that on case-insensitive FS like on OS X or Windows? Argh, I hate that also. :imp: 

> Of course, we can indicate in the doc, to format correctly the package name with your pattern, mais sérieusement, it is not legible, and the error message, if any, do not be understandable.

Alors, it's also very hard to debug - but maybe it also not obvious for the plugin and it would have to do too many checks.

## TL;dr

> IMO, I would tend to automatically include the repositories in the extra.asset-repositories, if possible.

So, composer just uses lowercase package names (with dash) and all package names have to be converted to be lowercase. That happens also for dependencies - within composer, not within the plugin(?). Therefore bower packages, which are case-sensitve may overlap.

Damn! Look impossible, indeed.

What's `proposal v2.0' btw?

**Comment by philippfrenzel** [comment]
My french although got pretty bad...:) nothing really constructive to add...

**Comment by francoispluchino** [comment]
@philippfrenzel Ok I just figured your message! Indeed, I allowed myself to write in [French](https://github.com/francoispluchino/composer-asset-plugin/issues/70#issuecomment-105929052)... Sorry, I didn't do proofreading (we'll put it on the back of tiredness ;-) )...

@schmunk42 

    How does bower handle that on case-insensitive FS like on OS X or Windows?

Bower installs the packages with names sensible to the case, but the file system overrides/merges the 2 contents. And there is no error message or warning.

    Alors, it's also very hard to debug - but maybe it also not obvious for the plugin and it would have to do too many checks

Yes.

    Therefore bower packages, which are case-sensitve may overlap.

Yes and no. The plugin can throw an exception if there is already a package name using a another repository url.

    What's `proposal v2.0' btw?

That is the issue #93. But it has 3 major disadvantages:
  - Use Node, So having to download the binaries if it is not installed, but it is not 100% sure that it works because of Dynamic Link Library (I tried on Linux and Windows and it is ok)
  - we are obliged to have GIT installed on the machine
  - but especially NPM and Bower does not work easily with private repositories, we need to do a hack in the config of GIT for redirected the HTTP urls to the GIT protocol, and use the SSH

There is an existing solution, but unfortunately - although I greatly respect their work - I'm not convinced by their approach. For me, IMO, if we use Node, then we must be able to use ALL the features of NPM and Bower; the plugin serving only to automate retrieval the information, and launch NPM and/or Bower. Another point I found, Bower does not automatically remove an installed package, if he is not present in the dependencies, we have to remove it manually, but it is a "normal" behavior , given that we can install dependencies without using the bower.json file.

**Comment by francoispluchino** [comment]
For NPM, now there is an enterprise version for use the private packages, see [NPM Enterprise](https://www.npmjs.com/enterprise). But I have not tested.

**Comment by schmunk42** [comment]
You could wrap the git, Node, Bower etc. in Docker ... but going that path is a big decision. I won't do it for now. 
And without the disadvantages of #93 are huge.

I think your plugin does a better job than bower itself ;)

Anyway, I think that ...

> The plugin can throw an exception if there is already a package name using a another repository url.

**would be great!**

And for the rest ... it is was it is, says the code. :) I'd rather like to speed-up the download-performance we discussed in another issue.

**Comment by alexandernst** [comment]
I ... have a really wild idea. Hold on, here it is. Use Reflection to hijack `strtolower` (and variants) in combination with `(debug_backtrace())[1]` so we can detect if  the caller function is the function that lowercases what we need and return the string as is.

Of course:

* This hijacking thing should be done the moment the plugin is loaded
* This is extremely hacky, but it's a solution and will work
* I know. It's hacky and ugly.

**Comment by alexandernst** [comment]
Oh, this actually requires RunKit which is not enabled (even installed) by default on PHP, so we're back to square 1. :disappointed:
